### PR TITLE
[FLINK-26771][Hive] Fix incomparable exception between boolean type and numeric type in Hive dialect

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To make Hive dialect support comparing boolean type with numeric/string type.

## Brief change log
  -  When it invovles with comparsion betwen boolean type and numeric/string type, delegate it to Hive's udf.

## Verifying this change
Added a  specifc sql statement in [HiveDialectQueryITCase.java](https://github.com/apache/flink/compare/master...luoyuxia:compare_boolean_int?expand=1#diff-fb5fc4106e27e27f9968e59b643003444fd40af58aa2f147ac301009b53a47ca)[#testAdditionalQueries](https://github.com/apache/flink/pull/19182/files#diff-fb5fc4106e27e27f9968e59b643003444fd40af58aa2f147ac301009b53a47caR159) to verify this fix. Without this fix, the test will fail.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by luoyuxia** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserSqlFunctionConverter.java [hunk 1]
```diff
@@ -415,24 +425,73 @@ public static SqlOperator getCalciteFn(
...
-                                    uInf.operandTypeInference,
-                                    uInf.operandTypeChecker,
-                                    SqlFunctionCategory.USER_DEFINED_FUNCTION,
-                                    deterministic);
+                // some functions should be handled as Hive UDF for
+                // the Hive specific logic
+                if (shouldHandleAsHiveUDF(uInf) && canHandledAsHiveUDF(uInf, functionConverter)) {
+                    calciteOp = asCalciteSqlFn(uInf, deterministic);
+                } else {
+                    calciteOp = HIVE_TO_CALCITE.get(hiveUdfName);
```
Comment (Myracle) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserSqlFunctionConverter.java:433]: A few minor suggestions:

1. The comments should be started with a capital letter.
2. shouldHandleAsHiveUDF->shouldHandledAsHiveUDF to be consistent with canHandledAsHiveUDF

Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserSqlFunctionConverter.java:433]: But about the comments, I think it's fine to not to change it for some some comments in this file are started with a capital letter and some are not.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**